In [119]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/datasets/sudalairajkumar/indian-startup-funding/startup_funding.csv


In [120]:
df= pd.read_csv ('/kaggle/input/datasets/sudalairajkumar/indian-startup-funding/startup_funding.csv')

In [121]:
df.head()

,Sr No,Date dd/mm/yyyy,Startup Name,Industry Vertical,SubVertical,City Location,Investors Name,InvestmentnType,Amount in USD,Remarks
0,1,09/01/2020,BYJU’S,E-Tech,E-learning,Bengaluru,Tiger Global Management,Private Equity Round,"20,00,00,000",NaN
1,2,13/01/2020,Shuttl,Transportation,App based shuttle service,Gurgaon,Susquehanna Growth Equity,Series C,"80,48,394",NaN
2,3,09/01/2020,Mamaearth,E-commerce,Retailer of baby and toddler products,Bengaluru,Sequoia Capital India,Series B,"1,83,58,860",NaN
3,4,02/01/2020,https://www.wealthbucket.in/,FinTech,Online Investment,New Delhi,Vinod Khatumal,Pre-series A,"30,00,000",NaN
4,5,02/01/2020,Fashor,Fashion and Apparel,Embroiled Clothes For Women,Mumbai,Sprout Venture Partners,Seed Round,"18,00,000",NaN


## **Inpecting the data**

In [122]:
# Shape of dataset
print("Rows:", df.shape[0], "Columns:", df.shape[1])

# Info
df.info()

# Missing values
df.isnull().sum()

Rows: 3044 Columns: 10
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3044 entries, 0 to 3043
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Sr No              3044 non-null   int64 
 1   Date dd/mm/yyyy    3044 non-null   object
 2   Startup Name       3044 non-null   object
 3   Industry Vertical  2873 non-null   object
 4   SubVertical        2108 non-null   object
 5   City  Location     2864 non-null   object
 6   Investors Name     3020 non-null   object
 7   InvestmentnType    3040 non-null   object
 8   Amount in USD      2084 non-null   object
 9   Remarks            419 non-null    object
dtypes: int64(1), object(9)
memory usage: 237.9+ KB


Sr No                   0
Date dd/mm/yyyy         0
Startup Name            0
Industry Vertical     171
SubVertical           936
City  Location        180
Investors Name         24
InvestmentnType         4
Amount in USD         960
Remarks              2625
dtype: int64

**Renaming Column Name**

In [123]:
df.rename(columns={
    'Sr No' : 'Sr_No',
    'Date dd/mm/yyyy': 'Date',
    'Startup Name' : 'StartUp',
    'Industry Vertical' : 'Vertical',
    'City  Location' : 'Location',
    'Investors Name' : 'Investors',
    'InvestmentnType' : 'Round',
    'Amount in USD' : 'Amount'
},inplace = True)

In [124]:
df.columns

Index(['Sr_No', 'Date', 'StartUp', 'Vertical', 'SubVertical', 'Location',
       'Investors', 'Round', 'Amount', 'Remarks'],
      dtype='object')

In [125]:
df.set_index('Sr_No')

,Date,StartUp,Vertical,SubVertical,Location,Investors,Round,Amount,Remarks
Sr_No,,,,,,,,,
1,09/01/2020,BYJU’S,E-Tech,E-learning,Bengaluru,Tiger Global Management,Private Equity Round,"20,00,00,000",NaN
2,13/01/2020,Shuttl,Transportation,App based shuttle service,Gurgaon,Susquehanna Growth Equity,Series C,"80,48,394",NaN
3,09/01/2020,Mamaearth,E-commerce,Retailer of baby and toddler products,Bengaluru,Sequoia Capital India,Series B,"1,83,58,860",NaN
4,02/01/2020,https://www.wealthbucket.in/,FinTech,Online Investment,New Delhi,Vinod Khatumal,Pre-series A,"30,00,000",NaN
5,02/01/2020,Fashor,Fashion and Apparel,Embroiled Clothes For Women,Mumbai,Sprout Venture Partners,Seed Round,"18,00,000",NaN
...,...,...,...,...,...,...,...,...,...
3040,29/01/2015,Printvenue,NaN,NaN,NaN,Asia Pacific Internet Group,Private Equity,"45,00,000",NaN
3041,29/01/2015,Graphene,NaN,NaN,NaN,KARSEMVEN Fund,Private Equity,"8,25,000",Govt backed VC Fund
3042,30/01/2015,Mad Street Den,NaN,NaN,NaN,"Exfinity Fund, GrowX Ventures.",Private Equity,"15,00,000",NaN


**Correcting the Data Type of Columns**

In [126]:
df['Date'] = pd.to_datetime(df['Date'], format='%d/%m/%Y', errors='coerce')
df['Amount'] = pd.to_numeric(df['Amount'].str.replace(',', ''), errors='coerce')


In [127]:
#df['Amount'] = df['Amount'].str.replace(',','').astype('int64')

- This ValueError occurs because your Amount column contains the text string 'undisclosed', which cannot be converted directly into an integer.
- To fix this, you should use pd.to_numeric() with the errors='coerce' parameter. This will automatically turn non-numeric text like "undisclosed" into NaN (null) values, allowing the rest of the conversion to proceed.
- With errors = 'coerce' If you run into a value that you can't convert, don't crash; just turn it into a NaN

### **Handling the Missing value**

In [128]:
df.isnull().sum()

Sr_No             0
Date              8
StartUp           0
Vertical        171
SubVertical     936
Location        180
Investors        24
Round             4
Amount          979
Remarks        2625
dtype: int64

- Whole row is null:  > df[df.isnull().all(axis=1)]
- At least one null in the row:  > df[df.isnull().any(axis=1)]

In [129]:
df[df.isnull().all(axis=1)]

,Sr_No,Date,StartUp,Vertical,SubVertical,Location,Investors,Round,Amount,Remarks


#### Fixing the inconsistency in the 'Amount' column -

In [130]:
df['Amount'].head()

0    200000000.0
1      8048394.0
2     18358860.0
3      3000000.0
4      1800000.0
Name: Amount, dtype: float64

**Change USD $ to INR (Rs)**

In [131]:
def to_inr(dollar):
    inr = dollar * 94.43  #Convert USD to INR
    return inr/10000000   #Converting INR to Carores

In [132]:
df['Amount']=df['Amount'].apply(to_inr)

In [133]:
df['Amount'].head()

0    1888.600000
1      76.000985
2     173.362715
3      28.329000
4      16.997400
Name: Amount, dtype: float64

In [134]:
#Filling the missing value in the Amount column 
mean= df['Amount'].mean()
median = df['Amount']. median()

print(f'Mean = {mean}')
print(f'Median = {median}')

Mean = 174.0335199282467
Median = 16.0531


Mean (174.03) is significantly higher than your Median (16.05). This usually means your 'Amount' column has outliers

In [135]:
# Calculate Q1 (25th percentile) and Q3 (75th percentile)
Q1 = df['Amount'].quantile(0.25)
Q3 = df['Amount'].quantile(0.75)
IQR = Q3 - Q1

# Define bounds (standard is 1.5 * IQR)
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter the DataFrame to remove outliers
df_cleaned = df[(df['Amount'] >= lower_bound) & (df['Amount'] <= upper_bound)]

print(f"Old shape: {df.shape}")
print(f"New shape: {df_cleaned.shape}")


Old shape: (3044, 10)
New shape: (1782, 10)


- Row count dropped from 3044 to 1782. This happened because the filter (df['Amount'] >= lower_bound) automatically excludes rows where the amount is NaN.

In [136]:
new_mean = df_cleaned['Amount'].mean()
new_median = df_cleaned['Amount'].median()

print(f'New Mean = {new_mean}')
print(f'New Median = {new_median}')


New Mean = 29.835263762530563
New Median = 9.443


Even after removing outliers, your Mean (29.8) is still about 3 times higher than your Median (9.44). This suggests your data is still "right-skewed

In [137]:
df.isnull().sum()

Sr_No             0
Date              8
StartUp           0
Vertical        171
SubVertical     936
Location        180
Investors        24
Round             4
Amount          979
Remarks        2625
dtype: int64

In [138]:
df['Amount'] = df['Amount'].fillna(new_median)

- Using the Median to fill 939 rows (about 30% of your data) is a common practice, but it comes with a trade-off you should consider.
- Filling nearly 1,000 rows with the exact same number (9.44) will create a huge "spike" in your data. This can make your correlations look stronger or weaker than they actually are because you've artificially reduced the variety in the 'Amount' column.

In [139]:
# Final clean-up after filling NaNs
df = df[(df['Amount'] >= lower_bound) & (df['Amount'] <= upper_bound)]


In [140]:
# Checking the misisng value percentage -
df.isnull().sum()/len(df) *100

Sr_No           0.000000
Date            0.289750
StartUp         0.000000
Vertical        5.541470
SubVertical    30.785947
Location        5.831221
Investors       0.833032
Round           0.108656
Amount          0.000000
Remarks        87.576965
dtype: float64

In [141]:
# Remark has more that 80 perc of missing value --> So we ca remove them 
df = df.drop(columns=['Remarks'])

In [142]:
df[df['Vertical'].isnull()]

,Sr_No,Date,StartUp,Vertical,SubVertical,Location,Investors,Round,Amount
2873,2874,2015-03-02,TrulyMadly.com,NaN,NaN,NaN,"Helion Venture Partners, Kae Capital",Private Equity,51.936500
2874,2875,2015-03-02,InstaLively,NaN,NaN,NaN,Group of Angel Investors,Seed Funding,1.133160
2875,2876,2015-03-03,Vserv,NaN,NaN,NaN,"Maverick Capital, IDG Ventures India",Private Equity,141.645000
2876,2877,2015-03-03,Intruo.com,NaN,NaN,NaN,"Ashutosh Lawania, Alok Goel & others",Seed Funding,9.443000
2877,2878,2015-03-05,Niffler,NaN,NaN,NaN,SAIF Partners,Seed Funding,9.443000
...,...,...,...,...,...,...,...,...,...
3039,3040,2015-01-29,Printvenue,NaN,NaN,NaN,Asia Pacific Internet Group,Private Equity,42.493500
3040,3041,2015-01-29,Graphene,NaN,NaN,NaN,KARSEMVEN Fund,Private Equity,7.790475
3041,3042,2015-01-30,Mad Street Den,NaN,NaN,NaN,"Exfinity Fund, GrowX Ventures.",Private Equity,14.164500
3042,3043,2015-01-30,Simplotel,NaN,NaN,NaN,MakeMyTrip,Private Equity,9.443000


In [143]:
df[df['Location'].isnull()]

,Sr_No,Date,StartUp,Vertical,SubVertical,Location,Investors,Round,Amount
338,339,2018-03-20,One Labs,Consumer Internet,Consumer Technology and Artificial Intelligenc...,NaN,Micromax Informatics Ltd,Private Equity,9.443000
1600,1601,2016-06-13,Matrubharti,Consumer Internet,e-Book Publisher,NaN,Viridian Capital,Seed Funding,0.283290
1765,1766,2016-04-12,Indi.com,Consumer Internet,Video Social Network,NaN,Anil Kapoor,Seed Funding,9.443000
1811,1812,2016-04-27,CareonGO,Consumer Internet,Online Pharmacy,NaN,"Farooq Oomerbhoy, Anupam Mittal, Anand Mittal,...",Seed Funding,9.443000
1891,1892,2016-03-23,HashLearn,Consumer Internet,Mobile based private tutoring services,NaN,"Bhavish Aggarwal, Binny Bansal,\\xc2\\xa0 Raja...",Seed Funding,9.443000
...,...,...,...,...,...,...,...,...,...
3039,3040,2015-01-29,Printvenue,NaN,NaN,NaN,Asia Pacific Internet Group,Private Equity,42.493500
3040,3041,2015-01-29,Graphene,NaN,NaN,NaN,KARSEMVEN Fund,Private Equity,7.790475
3041,3042,2015-01-30,Mad Street Den,NaN,NaN,NaN,"Exfinity Fund, GrowX Ventures.",Private Equity,14.164500
3042,3043,2015-01-30,Simplotel,NaN,NaN,NaN,MakeMyTrip,Private Equity,9.443000


In [144]:
df= df.dropna(subset=['Vertical','SubVertical','Location'], how= 'all')

In [145]:
# Filling the missing value
# 'SubVertical' (fillng it with unknown)
df['SubVertical'] = df['SubVertical'].fillna("Unknown")


In [146]:
#Filling the Investors columns
df['Investors'] = df['Investors'].fillna('Undisclosed')

In [147]:
# Checking the misisng value percentage -
df.isnull().sum()

Sr_No          0
Date           6
StartUp        0
Vertical       0
SubVertical    0
Location       8
Investors      0
Round          3
Amount         0
dtype: int64

In [148]:
df = df.dropna(subset = ['Date', 'StartUp', 'Vertical', 'Location', 'Investors', 'Round','Amount'])#

In [149]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2591 entries, 1 to 2872
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Sr_No        2591 non-null   int64         
 1   Date         2591 non-null   datetime64[ns]
 2   StartUp      2591 non-null   object        
 3   Vertical     2591 non-null   object        
 4   SubVertical  2591 non-null   object        
 5   Location     2591 non-null   object        
 6   Investors    2591 non-null   object        
 7   Round        2591 non-null   object        
 8   Amount       2591 non-null   float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(6)
memory usage: 202.4+ KB


In [150]:
#Checking for duplicate rows
duplicates = df.duplicated().sum()
print(f'The Dataframe has {duplicates} no. rows ')

The Dataframe has 0 no. rows 


#### **Fixing the inconsistency in the 'StartUp' column -**

In [151]:
#Getting all the startup name 
df['StartUp'].unique().shape 

(2181,)

In [152]:
df['StartUp'].unique().tolist()

['Shuttl',
 'Mamaearth',
 'https://www.wealthbucket.in/',
 'Fashor',
 'Pando',
 'Ecozen',
 'Healthians',
 'InCred',
 'Trell',
 'Misters',
 'Sunstone Eduversity Pvt. Ltd',
 'Burger Singh',
 'Aye Finance',
 'SuperGaming',
 'eBikeGo',
 'Medikabazaar',
 'The Man Company',
 'FPL Technologies',
 'Cashflo',
 'Digital F5',
 '3rdFlix',
 '75F',
 'Myelin Foundry',
 'Atomberg Technology',
 'Vyapar App',
 'Progcap',
 'MyPetrolPump',
 'Cars24',
 'Lo! Foods',
 'INDwealth',
 'HungerBox',
 'AdmitKard',
 'Mishry Reviews',
 'RenewBuy',
 'Atlan',
 'WizCounsel',
 'Ola Cabs',
 'Daalchini Technologies',
 'Ezyhaul',
 'Indus OS',
 'HealthAssure',
 'House of Msasaba',
 'Board Infinity',
 'Bira91',
 'Avail Finance',
 'Agara Labs',
 'Pratilipi',
 'Bolo App',
 'OkCredit',
 'Vogo Automotive Pvt. Ltd.',
 'Leegality',
 'Ola Electric',
 'Saahas Zero Waste',
 'StyleDotMe',
 'GlowRoad',
 'Medlife',
 'Kabadiwala',
 'Tripoto',
 'Azah',
 'Setu',
 'Toppr',
 'Craftsvilla',
 'Unacademy',
 'My Healthcare',
 'KrazyBee',
 'Incre

In [153]:
#By manual assessment we get the inconsistency in startup names ----> Resolving that 
df['StartUp'] = df['StartUp'].str.replace(r'\\\\xe2\\\\x80\\\\x99', "'", regex=True).str.replace(r'\\\\xe2\\\\x80\\\\x9d', '"', regex=True).str.replace(r'\\\\x[a-f0-9]{2}', '', regex=True).str.replace('https://www.wealthbucket.in/','Wealth Bucket').str.replace(r'\s+', ' ', regex=True).str.strip()
df['StartUp'] = df['StartUp'].str.title().str.strip()

In [154]:
df['StartUp'].unique().tolist()

['Shuttl',
 'Mamaearth',
 'Wealth Bucket',
 'Fashor',
 'Pando',
 'Ecozen',
 'Healthians',
 'Incred',
 'Trell',
 'Misters',
 'Sunstone Eduversity Pvt. Ltd',
 'Burger Singh',
 'Aye Finance',
 'Supergaming',
 'Ebikego',
 'Medikabazaar',
 'The Man Company',
 'Fpl Technologies',
 'Cashflo',
 'Digital F5',
 '3Rdflix',
 '75F',
 'Myelin Foundry',
 'Atomberg Technology',
 'Vyapar App',
 'Progcap',
 'Mypetrolpump',
 'Cars24',
 'Lo! Foods',
 'Indwealth',
 'Hungerbox',
 'Admitkard',
 'Mishry Reviews',
 'Renewbuy',
 'Atlan',
 'Wizcounsel',
 'Ola Cabs',
 'Daalchini Technologies',
 'Ezyhaul',
 'Indus Os',
 'Healthassure',
 'House Of Msasaba',
 'Board Infinity',
 'Bira91',
 'Avail Finance',
 'Agara Labs',
 'Pratilipi',
 'Bolo App',
 'Okcredit',
 'Vogo Automotive Pvt. Ltd.',
 'Leegality',
 'Ola Electric',
 'Saahas Zero Waste',
 'Styledotme',
 'Glowroad',
 'Medlife',
 'Kabadiwala',
 'Tripoto',
 'Azah',
 'Setu',
 'Toppr',
 'Craftsvilla',
 'Unacademy',
 'My Healthcare',
 'Krazybee',
 'Increff',
 'Fleetx',

#### **Fixing the inconsistency in the 'Vertical' column -**

In [155]:
df['Vertical'].unique().tolist()

['Transportation',
 'E-commerce',
 'FinTech',
 'Fashion and Apparel',
 'Logistics',
 'Technology',
 'B2B-focused foodtech startup',
 'Finance',
 'Video',
 'Health and wellness',
 'Education',
 'Food and Beverage',
 'Health and Wellness',
 'Video Games',
 'Last Mile Transportation',
 'Healthcare',
 'Consumer Goods',
 'Advertising, Marketing',
 'SaaS',
 'IoT',
 'Information Technology',
 'Consumer Technology',
 'Accounting',
 'Retail',
 'E-Commerce',
 'EdTech',
 'Services',
 'Compliance',
 'Transport',
 'Tech',
 'Health Care',
 'Luxury Label',
 'Deep-Tech',
 'Digital Media',
 'Waste Management Service',
 'Social Media',
 'Fintech',
 'Edtech',
 'AI',
 'Nanotechnology',
 'Travel Tech',
 'Online Education',
 'Online Marketplace',
 'B2B',
 'SaaS, Ecommerce',
 'Gaming',
 'Food',
 'Food Tech',
 'Social Network',
 'Financial Tech',
 'Fashion',
 'Automobile',
 'Real Estate',
 'Logistics Tech',
 'Consumer Internet',
 'Fin-Tech',
 'Food-Tech',
 'Ed-Tech',
 'IT',
 'Ed-tech',
 'Clean-tech',
 'Fiinan

In [156]:
#By manual assessment we get the inconsistency in vertical names ----> Resolving that 
#Fix the 'é' (like in Café or Décor)
df['Vertical'] = df['Vertical'].str.replace(r'\\\\xc3\\\\xa9', 'e', regex=True)

# 2. Fix the apostrophe (like in Byju's or Caf's)
df['Vertical'] = df['Vertical'].str.replace(r'\\\\xe2\\\\x80\\\\x99', "'", regex=True)

# 3. Wipe out the broken space junk (\\\\xc2\\\\xa0)
df['Vertical'] = df['Vertical'].str.replace(r'\\\\xc2\\\\xa0', ' ', regex=True)

# 4. Fix Newlines (\\\\Nplatform) found in 'Mobile Growth Hacking'
df['Vertical'] = df['Vertical'].str.replace(r'\\\\n', ' ', regex=True, case=False)

# 5. Final Polish: Fix casing and remove extra whitespace
df['Vertical'] = df['Vertical'].str.strip().str.title()

# 6. Optional: Clean up repetitive suffixes to help grouping
# This merges "Booking Platform" and "Booking App" into just "Booking"
df['Vertical'] = df['Vertical'].str.replace(r'\s+(Platform|App|Marketplace|Service|Solution)s?\b', '', regex=True, case=False)

In [157]:
df['Vertical'].unique().tolist()

['Transportation',
 'E-Commerce',
 'Fintech',
 'Fashion And Apparel',
 'Logistics',
 'Technology',
 'B2B-Focused Foodtech Startup',
 'Finance',
 'Video',
 'Health And Wellness',
 'Education',
 'Food And Beverage',
 'Video Games',
 'Last Mile Transportation',
 'Healthcare',
 'Consumer Goods',
 'Advertising, Marketing',
 'Saas',
 'Iot',
 'Information Technology',
 'Consumer Technology',
 'Accounting',
 'Retail',
 'Edtech',
 'Services',
 'Compliance',
 'Transport',
 'Tech',
 'Health Care',
 'Luxury Label',
 'Deep-Tech',
 'Digital Media',
 'Waste Management',
 'Social Media',
 'Ai',
 'Nanotechnology',
 'Travel Tech',
 'Online Education',
 'Online',
 'B2B',
 'Saas, Ecommerce',
 'Gaming',
 'Food',
 'Food Tech',
 'Social Network',
 'Financial Tech',
 'Fashion',
 'Automobile',
 'Real Estate',
 'Logistics Tech',
 'Consumer Internet',
 'Fin-Tech',
 'Food-Tech',
 'Ed-Tech',
 'It',
 'Clean-Tech',
 'Fiinance',
 'Food & Beverages',
 'Media',
 'Publishing',
 'Food And Beverages',
 'Agriculture',
 'En

#### **Fixing the inconsistency in the 'SubVertical' column -**

In [158]:
df['SubVertical'].unique().tolist()

['App based shuttle service',
 'Retailer of baby and toddler products',
 'Online Investment',
 'Embroiled Clothes For Women',
 'Open-market, freight management platform',
 'Agritech',
 'Food Solutions For Corporate',
 'Non-Banking Financial Company',
 'Experience Discovery Platform',
 "Men's Health and Wellness brand",
 'Elearning',
 'Indian Burger Brand',
 'Healthcare services',
 'Financial Services To MSMEs',
 'Social gaming platform',
 'Electric bike rental',
 'B2B platform for medical supplies',
 'Beauty and Grooming',
 'Financial Services',
 'Invoice discounting platform and SME lending marketplace',
 'Digital marketing firm',
 'Education Technology',
 'Building automation system',
 'Deep-technology',
 'Consumer Electronics, Home Appliances',
 'Mobile-based Accounting Software',
 'Supply Chain Management',
 'Fuel Delivery',
 'Car Retail',
 'Low carb food for Diabetics',
 'Wealth Management',
 'B2B Foodtech',
 'University Admissions',
 'Product Review',
 'Auto Insurance',
 'Big Dat

In [159]:
# 1. Fix the broken apostrophe (\\\\xe2\\\\x80\\\\x99)
df['SubVertical'] = df['SubVertical'].str.replace(r'\\\\xe2\\\\x80\\\\x99', "'", regex=True)

# 2. Fix the broken space junk (\\\\xc2\\\\xa0) from previous examples
df['SubVertical'] = df['SubVertical'].str.replace(r'\\\\xc2\\\\xa0', ' ', regex=True)

# 3. Standardize casing and trim extra whitespace
df['SubVertical'] = df['SubVertical'].str.title().str.strip()

In [160]:
df['SubVertical'].unique().tolist()

['App Based Shuttle Service',
 'Retailer Of Baby And Toddler Products',
 'Online Investment',
 'Embroiled Clothes For Women',
 'Open-Market, Freight Management Platform',
 'Agritech',
 'Food Solutions For Corporate',
 'Non-Banking Financial Company',
 'Experience Discovery Platform',
 "Men'S Health And Wellness Brand",
 'Elearning',
 'Indian Burger Brand',
 'Healthcare Services',
 'Financial Services To Msmes',
 'Social Gaming Platform',
 'Electric Bike Rental',
 'B2B Platform For Medical Supplies',
 'Beauty And Grooming',
 'Financial Services',
 'Invoice Discounting Platform And Sme Lending Marketplace',
 'Digital Marketing Firm',
 'Education Technology',
 'Building Automation System',
 'Deep-Technology',
 'Consumer Electronics, Home Appliances',
 'Mobile-Based Accounting Software',
 'Supply Chain Management',
 'Fuel Delivery',
 'Car Retail',
 'Low Carb Food For Diabetics',
 'Wealth Management',
 'B2B Foodtech',
 'University Admissions',
 'Product Review',
 'Auto Insurance',
 'Big Dat

In [ ]:
#### **Fixing the inconsistency in the 'Round' column -**

In [ ]:
df['Round'].unique().tolist()

In [162]:
  Mapping={
    # Pre-Series A
    'Pre-Series A': 'Pre-Series A',
    'Pre Series A': 'Pre-Series A',
    'Pre-Series A ': 'Pre-Series A',

    # Debt
    'Debt Funding': 'Debt',
    'Debt-Funding': 'Debt'}

In [166]:
df['Round']=df['Round'].replace('Mapping')

/tmp/ipykernel_57/2861844759.py:1: FutureWarning: Series.replace without 'value' and with non-dict-like 'to_replace' is deprecated and will raise in a future version. Explicitly specify the new values instead.
  df['Round']=df['Round'].replace('Mapping')


In [167]:
df['Round'].str.replace('Funding','').str.replace('Round','').str.replace('Equity','').str.strip().str.title().unique()

array(['Series C', 'Series B', 'Pre-Series A', 'Seed', 'Series A', 'Debt',
       'Venture', 'Corporate', 'Maiden', 'Private', 'Series D', 'Angel',
       'Series J', 'Venture - Series Unknown', 'Bridge',
       'Debt And Preference Capital', 'Inhouse', 'Seed/ Angel',
       'Pre Series A', 'Debt-', '', 'Mezzanine', 'Series B (Extension)',
       'Series E', 'Seed / Angel', 'Seed/Angel', 'Seed Funding',
       'Seed / Angle', 'Angel / Seed', 'Structured Debt', 'Term Loan'],
      dtype=object)

#### **Fixing the inconsistency in the 'Investors' column -**

In [ ]:
df['Investors'].unique().tolist()

I am using a dual-dataframe approach to handle rows with multiple Investors name 
The dataset contains multiple investor names in a single row separated by commas.  
To perform proper investor-level analysis, a separate dataframe was created.

- `df_original` is used for:
  - Funding analysis
  - Sector analysis
  - City analysis
  - Startup analysis

- `df_investor` is used for:
  - Investor analysis
  - Top investors
  - Investor participation trends



In [ ]:
df['Investors']=df['Investors'].str.replace('\\\\xe2\\\\x80\\\\x99s','').str.replace('\\\\xc2\\\\xa0','').str.replace('and','')
df['Investors']=df['Investors'].str.title().str.strip()

In [ ]:
df_by_investor = df.copy()
df_by_investor['Investors'] = df_by_investor['Investors'].str.split(',')
df_by_investor= df_by_investor.explode('Investors')

In [ ]:
df['Investors'].str.split(',').explode()

* The `explode()` function was applied on the investor column to separate multiple investors into individual rows for better analysis and visualization.
* The `explode()` function is used to transform list-like values in a column into separate rows.


In [ ]:
df_by_investor.sample(5)

In [ ]:
#Cleaning
df_by_investor['Investors'] = df_by_investor['Investors'].str.replace(',','').str.replace(',','').str.title().str.strip()

#### **Fixing the inconsistency in the 'Location' column -**

In [ ]:
df['Location'].unique()

In [ ]:
#Clean up known messy encoding/whitespace
df['Location'].str.replace(r'\\\\xc2\\\\xa0','',regex=True).unique()

In [ ]:
df['Location'] = df['Location'].str.replace(r'\\\\xc2\\\\xa0','',regex=True).str.strip()

In [ ]:
#Standardize delimiters to a single '/'
df['Location'] = df['Location'].replace(r'\s*(/|&|and|,|&)\s*', '/', regex=True)

In [ ]:
df['Location'].unique()

I am using a dual-dataframe approach to handle rows with multiple cities (e.g., "Mumbai / Bangalore").

**1. Why am I doing this?**
*   **To get accurate city counts:** A startup listed as "Mumbai / Bangalore" won't show up in a search for just "Mumbai." By splitting them, I can see the true activity in every city.
*   **To avoid "Money Math" errors:** Using `explode()` duplicates the row. If I'm not careful, a $10M investment will be counted twice ($10M for Mumbai + $10M for Bangalore = $20M). 

**2. The Solution:**
*   **`df` (Original):** I will use this for all **Financial Analysis** (Total/Average Funding) to ensure the dollar amounts are 100% accurate.
*   **`df_city` (Exploded):** I will use this copy for **Geographic Analysis** (City counts and maps) so that every location is represented.


In [ ]:
# Create a dedicated copy for geographic analysis
df_by_city = df.copy()


In [ ]:
#Split and Explode
df_by_city['Location'] = df_by_city['Location'].str.split('/')
df_by_city = df_by_city.explode('Location')
df_by_city['Location'].unique()

**Standardize the Location name**

In [ ]:
city_mapping = {
    # Spelling Fixes & Standardizing to common names
    'Bengaluru': 'Bangalore',
    'Gurugram': 'Gurgaon',
    'Kolkatta': 'Kolkata',
    'Ahemadabad': 'Ahmedabad',
    'Ahemdabad': 'Ahmedabad',
    'Nw Delhi': 'New Delhi',
    'Delhi': 'New Delhi',
    'Kormangala': 'Bangalore', # Neighborhood in Bangalore
    'Andheri': 'Mumbai',       # Neighborhood in Mumbai
    'Chembur': 'Mumbai',       # Neighborhood in Mumbai
    
    # State names mapped to their primary startup hubs 
    'Haryana': 'Gurgaon',
    'Karnataka': 'Bangalore',
    'Kerala': 'Kochi',
    'Uttar Pradesh': 'Noida',
    
    # Cleaning up common International/Combined labels
    'USA': 'US',
    'SFO': 'San Francisco',
    'NY': 'New York'
}

In [ ]:
df_by_city['Location'] = df_by_city['Location'].replace(city_mapping)
df_by_city['Location'].unique()

In [ ]:
df['Location'] = df['Location'].replace(city_mapping)
df['Location'].unique()